code/DynamicCache.php

```diff
@@ -384,6 +393,10 @@ public function run($url)
         // This prevents a new user's security token from being regenerated incorrectly
         $_SESSION['SecurityID'] = SecurityToken::getSecurityID();
 
+        // Set the stage of the website
+        // This is normally called in VersionedRequestFilter.
+        Versioned::choose_site_stage();
```

Comment:
Ok, I've read all the code, and it looks like you'll end up with two session objects, which I was worried about, but as long as bootstrapping is performed on both equivalently, it is probably fine...

My second point stands, however; There is a lack of permission checking on can_choose_site_stage().

If you use dynamiccache, an unauthenticated user can still view draft content, which is a huge risk.

Can you double check to see if this is the case?